maltoolbox/language/languagegraph.py

```diff
@@ -1145,21 +1664,21 @@ def get_association_by_fields_and_assets(
         The association matching the fieldnames and asset types.
         None if there is no match.
         """
-        first_asset = self.get_asset_by_name(first_asset_name)
+        first_asset = self.assets[first_asset_name]
         if first_asset is None:
             raise LookupError(
                 f'Failed to find asset with name \"{first_asset_name}\" in '
                 'the language graph.'
             )
 
-        second_asset = self.get_asset_by_name(second_asset_name)
+        second_asset = self.assets[second_asset_name]
         if second_asset is None:
             raise LookupError(
                 f'Failed to find asset with name \"{second_asset_name}\" in '
```

Comment:
The discussion here is not specific to this line. It is for wherever `self.get_asset_by_name` was removed.

#### First issue:

Is there a case now where `self.assets[key]` will return `None`? Aren't assets added to the self.assets dict when they are ready objects?

#### Exception handling:

Also, from python's side, `LookupError` is a base class. A child class called `KeyError` is what should be used instead for non-existent key lookup errors (and IndexError for non-existent list keys). The code could look like:

```
try:
    second_asset = self.assets[second_asset_name]
except KeyError as e:
    raise LanggraphError("Failed to find asset with name %s in the langgraph...", second_asset_name) from e
```

In the snippet, notice:

- the use of the %s formatter instead of f-strings
- the use of a custom exception class

If we don't want to use the custom exception class, then it is fine to not handle the exception; there is no benefit in capturing the exception just to raise it again.

#### Utility of exception handling

But there is also the question, do we expect this error to happen? Do we expect `second_asset_name` to not exist in `self.assets`? This is the language graph, so if this happens, this is a bug in the mal file...........

#### But before we discuss this more:

This function is called only by translators and ingestors. I suggest that for the rewrite we drop support for those. The neo4j ingestor uses a library that is EOL anyway. Securicad model files can be translated to json (I think). So, these two modules are not strictly needed. When the rewrite is done, we can see how we support these if we still need them and in the mean time the lang graph will have a more focused scope. For example, this function does not handle properly cases where two associations share the same field and asset names. Instead of fixing this, let's focus on the needs of the core toolbox and simulator and I am sure ingestors and translators will be able to be supported without issues and further custom additions like this function.

Reply:
> The discussion here is not specific to this line. It is for wherever `self.get_asset_by_name` was removed.
> #### First issue:
> 
> Is there a case now where `self.assets[key]` will return `None`? Aren't assets added to the self.assets dict when they are ready objects?

I don't think so and I also think that I've been doing too much error handling for some of these methods.

> #### Exception handling:
> 
> Also, from python's side, `LookupError` is a base class. A child class called `KeyError` is what should be used instead for non-existent key lookup errors (and IndexError for non-existent list keys). The code could look like:
> 
> ```
> try:
>     second_asset = self.assets[second_asset_name]
> except KeyError as e:
>     raise LanggraphError("Failed to find asset with name %s in the langgraph...", second_asset_name) from e
> ```
> 
> In the snippet, notice:
> 
>     * the use of the %s formatter instead of f-strings
> 
>     * the use of a custom exception class
> 
> 
> If we don't want to use the custom exception class, then it is fine to not handle the exception; there is no benefit in capturing the exception just to raise it again.
> #### Utility of exception handling
> 
> But there is also the question, do we expect this error to happen? Do we expect `second_asset_name` to not exist in `self.assets`? This is the language graph, so if this happens, this is a bug in the mal file...........

I agree with you that I think we only want to do extra exception handling if we can add anything of values to the debugging of the exception. I am not sure that it particularly helps all that much in this case. You basically know that a key is missing from the list of assets even with the default behaviour.


> #### But before we discuss this more:
> 
> This function is called only by translators and ingestors. I suggest that for the rewrite we drop support for those. The neo4j ingestor uses a library that is EOL anyway. Securicad model files can be translated to json (I think). So, these two modules are not strictly needed. When the rewrite is done, we can see how we support these if we still need them and in the mean time the lang graph will have a more focused scope.

This is something the rewrite does impact, it uses the association full names to identify them, which is `<association name>_<left_field_name>_<left_field_asset_name>_<right_field_name>_<right_field_asset_name>`. And this should always be a unique identifier so we can use this more reliably down the line. I supposed you technically could have an asset with that exact name, but even if the compiler doesn't reject it we should reject it in the mal-toolbox.

> For example, this function does not handle properly cases where two associations share the same field and asset names.

Nothing could handle it based on field names and asset names alone.

> Instead of fixing this, let's focus on the needs of the core toolbox and simulator and I am sure ingestors and translators will be able to be supported without issues and further custom additions like this function.

The ingestors and translators are out of date right now anyway. So, the focus has not been on them, it is simply that I did not check where all of these functions are used. I mostly focused on the behaviour I wanted so far, but I do think it is a useful component of the upcoming review.